### PR TITLE
[jamendo] Add new extractor

### LINKED ISSUE #10132: Support downloads from jamendo.com
Jamendo is a platform that aggregates music released under Creative Commons license. That would be nice if you guys could support downloads from there.

Here is the example of a track URL: https://www.jamendo.com/track/196219/stories-from-emona-i.

### PR DESCRIPTION
**Pull request by Thor77**
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [X] New extractor

---

### Description of your *pull request* and other information

Adding an extractor for [Jamendo](https://www.jamendo.com/)

Fix #10132

### CODE REVIEW

File: youtube_dl/extractor/jamendo.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    }
+
+    def _real_extract(self, url):
+        url_data = self._VALID_URL_RE.match(url)
+        track_id = url_data.group('id')
+        webpage = self._download_webpage(url, track_id)
+
+        thumbnail = self._html_search_regex(r'<meta itemprop="image" content="(.+)">', webpage, 'thumbnail', fatal=False)
+        title = self._html_search_regex('<meta itemprop="name" content="(.+)">', webpage, 'title')
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Use `_html_search_meta` instead.

File: youtube_dl/extractor/jamendo.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        return {
+            'id': track_id,
+            'display_id': url_data.group('display_id'),
+            'ext': 'mp3',
+            'title': title,
+            'thumbnail': thumbnail,
+            'url': 'https://mp3d.jamendo.com/download/track/%s/mp32' % track_id
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: ogg is also available.

File: youtube_dl/extractor/jamendo.py (hunk 1 above)
Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Oh, sorry. Due to the lack of proper documentation for that function I couldn't really figure out what it's being used for

File: youtube_dl/extractor/jamendo.py (hunk 2 above)
Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: track/<id>/ogg just redirects to /mp32

Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Wrong. https://ogg.jamendo.com/?trackid=677050&format=ogg1&from=app-97dab294

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Oh, you're right. Will add that, thanks 👍

Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Also where did you get `https://mp3d.jamendo.com/download/track/%s/mp32` from? Browser uses different location.

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: It's the source of files you receive after clicking "Download".
https://ogg.jamendo.com/download/track/677050/ogg is also available.
I would like to choose one url-format for both, not sure which one, though.

Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: https://mp3l.jamendo.com/?trackid=196219&format=mp31&from=app-97dab294

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Yeah, I know. Both formats are available through both url-formats.
Which one do you want me to use?

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Hm, the official API is redirecting to `https://mp3d.jamendo.com/download/track/677050/mp32/` for `https://api.jamendo.com/v3.0/tracks/file/?client_id=56d30c95&audioformat=mp32&id=677050` (and similar ones for `flac`, `ogg` and `mp31`)

Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: I've already pointed out: API URLs should be used.

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Uh, I interpreted that as "URLs the API is using", I'm sorry. Additionally only mp3 and ogg seem to be available via that method:
```json
{
    "stream": {
        "mp3": "https://mp3l.jamendo.com/?trackid=196219&format=mp31&from=app-97dab294&u=3118107",
        "ogg": "https://ogg.jamendo.com/?trackid=196219&format=ogg1&from=app-97dab294&u=3118107"
    },
    "download": {
        "mp3": "https://mp3d.jamendo.com/download/track/196219/mp32/?u=3118107",
        "ogg": "https://ogg.jamendo.com/download/track/196219/ogg1/?u=3118107"
    }
}
```
On the upside I could easily extract them from the sites source-code and don't have to hope Jamendo doesn't change them in the future (but that could also hapen to this data-dict...).
Is that the reason you prefer them?
I personally would prefer more formats, but that decision is obviously up to you.

Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: flac is available just fine https://flac.jamendo.com/?trackid=677050&format=flac&from=app-97dab294.

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: But not from the site source code, is it?

File: youtube_dl/extractor/jamendo.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        track_id = url_data.group('id')
+        webpage = self._download_webpage(url, track_id)
+
+        thumbnail = self._html_search_meta(
+            'image', webpage, 'thumbnail', fatal=False)
+        title = self._html_search_meta('name', webpage, 'title')
+        formats = [
+            {
+                'format_id': 'mp31',
+                'url': 'https://mp3l.jamendo.com/?trackid=%s&format=mp31'
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: `from=app-97dab294` should be preserved.

File: youtube_dl/extractor/jamendo.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                'url': 'https://ogg.jamendo.com/?trackid=%s&format=ogg1'
+                % track_id,
+                'ext': 'ogg'
+            },
+            {
+                'format_id': 'flac',
+                'url': 'https://flac.jamendo.com/?trackid=%s&format=flac'
+                % track_id,
+                'ext': 'flac'
+            }
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Code duplication should be eliminated.

File: youtube_dl/extractor/jamendo.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    _TEST = {
+        'url': 'https://www.jamendo.com/track/196219/stories-from-emona-i',
+        'md5': '6e9e82ed6db98678f171c25a8ed09ffd',
+        'info_dict': {
+            'id': '196219',
+            'display_id': 'stories-from-emona-i',
+            'ext': 'flac',
+            'title': 'Stories from Emona I',
+            'thumbnail': 're:^https?://.*\.jpg',
+            'url': 'https://flac.jamendo.com/?trackid=196219&format=flac'
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: There is no need to test URLs.

File: youtube_dl/extractor/jamendo.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                'format_id': 'ogg',
+                'url': 'https://ogg.jamendo.com/?trackid=%s&format=ogg1'
+                % track_id,
+                'ext': 'ogg'
+            },
+            {
+                'format_id': 'flac',
+                'url': 'https://flac.jamendo.com/?trackid=%s&format=flac'
+                % track_id,
+                'ext': 'flac'
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Formats should have `quality`.

File: youtube_dl/extractor/jamendo.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            'title': 'Stories from Emona I',
+            'thumbnail': 're:^https?://.*\.jpg',
+            'url': 'https://flac.jamendo.com/?trackid=196219&format=flac'
+        }
+    }
+
+    def _real_extract(self, url):
+        url_data = self._VALID_URL_RE.match(url)
+        track_id = url_data.group('id')
+        webpage = self._download_webpage(url, track_id)
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: `display_id` should be used for console output.

File: youtube_dl/extractor/jamendo.py (hunk 4 above)
Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Done. Hope it's okey to use a `namedtuple` here.

File: youtube_dl/extractor/jamendo.py [hunk 8]
```diff
@@ -17,49 +21,44 @@ class JamendoIE(InfoExtractor):
...
+        url_template = 'https://%s.jamendo.com/?trackid=%s&format=%s&from=app-97dab294'
+        format_data = [
+            FormatData(
+                format_id='mp31', sub_domain='mp3l', ext='mp3', quality=0),
+            FormatData(
+                format_id='mp32', sub_domain='mp3d', ext='mp3', quality=1),
+            FormatData(
+                format_id='ogg1', sub_domain='ogg', ext='ogg', quality=2),
+            FormatData(
+                format_id='flac', sub_domain='flac', ext='flac', quality=3),
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Still a code duplication.

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: using `'flac'` three times?

Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Using `FormatData` four times.

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: I don't understand your concern.
Do you want me to use this just to not use the word `FormatData` four times?
```python
[
    FormatData(format_id, sub_domain, ext, quality)
    for format_id, sub_domain, ext, quality in [
        ('mp31', 'mp3l', 'mp3', 0),
        ('mp32', 'mp3d', 'mp3', 1),
        ('ogg1', 'ogg', 'ogg', 2),
        ('flac', 'flac', 'flac', 3),
    ]
]
```

File: youtube_dl/extractor/jamendo.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        formats = [
+            {
+                'format_id': fd.format_id,
+                'url': url_template % (fd.sub_domain, track_id, fd.format_id),
+                'ext': fd.ext,
+                'quality': fd.quality
+            }
+            for fd in format_data
+        ]
+        self._check_formats(formats, video_id=display_id)
```
Comment (dstftw) [inline youtube_dl/extractor/jamendo.py]: Example URL this is necessary for?

Comment (Thor77) [inline youtube_dl/extractor/jamendo.py]: Nope, but I'm not really sure if all formats are available in all formats, though.

**Comment by Thor77** [comment]
Thanks :)

**Comment by dstftw** [comment]
Do you want to be credited as one of the authors of youtube-dl? If so under what name?

**Comment by Thor77** [comment]
Thanks for the offer, I'd like to. Just use my nickname, please: `Thor77`

**Comment by FoilBlue** [comment]
Heres a code dump of a bash script i was writing for downloading flacs from jamendo. I needed more work but theres code in there for adding flac/ogg metadata tags and api urls and json tags for the metadata fields. Hope it is help! :) thanks for adding much needed support to ytdl :D.
http://software.aross.me/jamendo_flac_downloader/indev_jamendo-dl_api3/

**Comment by Thor77** [comment]
@FoilBlue Please consider creating a own Pull Request, as it would take everyone else more time to understand your code and then port it into this extractor.

**Pull request merged**